src/multicalendarstorage.cpp

```diff
@@ -0,0 +1,532 @@
+}
+
+bool MultiCalendarStorage::setTimeZone(const QTimeZone &timezone)
+{
+    bool changed = false;
+    d->mTimezone = timezone;
+    for (QHash<QString, CalendarHandler>::Iterator it = d->mCalendars.begin();
+         it != d->mCalendars.end(); it++) {
+        changed = changed || (it->calendar()->timeZone() != timezone);
+        it->calendar()->setTimeZone(timezone);
+    }
+    return changed;
+}
+
+bool MultiCalendarStorage::open()
+{
+    for (const Notebook::Ptr &notebook : loadedNotebooks(&d->mDefaultNotebookUid)) {
+        QHash<QString, CalendarHandler>::Iterator it = d->mCalendars.find(notebook->uid());
+        if (it == d->mCalendars.end()) {
+            it = d->mCalendars.insert(notebook->uid(), CalendarHandler(d->mTimezone));
+        }
+        it->setNotebook(notebook);
+    }
+    return true;
+}
+
+bool MultiCalendarStorage::close()
+{
+    d->mCalendars.clear();
+    return true;
+}
+
+QString MultiCalendarStorage::instanceIdentifier(const QString &notebookUid,
+                                                 const Incidence &incidence)
+{
+    return instanceIdentifier(notebookUid, incidence.instanceIdentifier());
+}
+
+QString MultiCalendarStorage::instanceIdentifier(const QString &notebookUid,
+                                                 const QString &identifier)
+{
+    return QString::fromLatin1("%1::NBUID::%2").arg(notebookUid).arg(identifier);
+}
+
+static QPair<QString, QString> deserialiseIdentifier(const QString &instanceIdentifier)
+{
+    int at = instanceIdentifier.indexOf(QString::fromLatin1("::NBUID::"));
+    return at < 0 ? QPair<QString, QString>()
+                  : QPair<QString, QString>(instanceIdentifier.left(at), instanceIdentifier.right(at + 9));
+}
+
+bool MultiCalendarStorage::loadIncidenceInstance(const QString &instanceIdentifier)
+{
+    QPair<QString, QString> ids = deserialiseIdentifier(instanceIdentifier);
+    const QString notebookUid = ids.first;
+    QString uid = ids.second;
+    if (notebookUid.isEmpty() || uid.isEmpty()) {
+        qCWarning(lcMkcal) << "wrong instance identifier" << instanceIdentifier;
+        return false;
+    }
+
+    // At the moment, from KCalendarCore, if the instance is an exception,
+    // the instanceIdentifier will ends with yyyy-MM-ddTHH:mm:ss[Z|[+|-]HH:mm]
+    // This is tested in tst_loadIncidenceInstance() to ensure that any
+    // future breakage would be properly detected.
+    if (uid.endsWith('Z')) {
+        uid = uid.left(uid.length() - 20);
+    } else if (uid.length() > 19 && uid[uid.length() - 9] == 'T') {
+        uid = uid.left(uid.length() - 19);
+    } else if (uid.length() > 25 && uid[uid.length() - 3] == ':') {
+        uid = uid.left(uid.length() - 25);
+    }
+
+    // Even if we're looking for a specific incidence instance, we load all
+    // the series for recurring event, to avoid orphaned exceptions in the
+    // calendar or recurring events without their exceptions.
+    QHash<QString, Incidence::List> hash;
```

Comment:
Quite minor: could restrict scope by moving into else branch. Up to you.

Reply:
Ah yes, indeed. Done.